libs/isograph-react/src/loadable-hooks/usePagination.ts

```diff
@@ -0,0 +1,315 @@
+import { ItemCleanupPair } from '@isograph/disposable-types';
+import {
+  UNASSIGNED_STATE,
+  useUpdatableDisposableState,
+} from '@isograph/react-disposable-state';
+import {
+  createReferenceCountedPointer,
+  ReferenceCountedPointer,
+} from '@isograph/reference-counted-pointer';
+import { useState } from 'react';
+import { subscribeToAnyChange } from '../core/cache';
+import { FragmentReference } from '../core/FragmentReference';
+import { getPromiseState, readPromise } from '../core/PromiseWrapper';
+import {
+  readButDoNotEvaluate,
+  type WithEncounteredRecords,
+} from '../core/read';
+import { LoadableField, type ReaderAst } from '../core/reader';
+import { useIsographEnvironment } from '../react/IsographEnvironmentProvider';
+import { useSubscribeToMultiple } from '../react/useReadAndSubscribe';
+import { maybeUnwrapNetworkRequest } from '../react/useResult';
+
+type FirstOrAfter = 'first' | 'after';
+type OmitFirstAfter<TArgs> = keyof Omit<TArgs, FirstOrAfter> extends never
+  ? void | Record<string, never>
+  : Omit<TArgs, FirstOrAfter>;
+
+type UsePaginationReturnValue<
+  TReadFromStore extends { parameters: object; data: object },
+  TItem,
+  TArgs,
+> =
+  | {
+      kind: 'Pending';
+      pendingFragment: FragmentReference<TReadFromStore, ReadonlyArray<TItem>>;
+      results: ReadonlyArray<TItem>;
+    }
+  | {
+      kind: 'Complete';
+      fetchMore: (args: OmitFirstAfter<TArgs>, first: number) => void;
+      results: ReadonlyArray<TItem>;
+      hasNextPage: boolean;
+    };
+
+type LoadedFragmentReferences<
+  TReadFromStore extends { parameters: object; data: object },
+  TItem,
+> = ReadonlyArray<LoadedFragmentReference<TReadFromStore, TItem>>;
+
+type LoadedFragmentReference<
+  TReadFromStore extends { parameters: object; data: object },
+  TItem,
+> = ItemCleanupPair<
+  ReferenceCountedPointer<FragmentReference<TReadFromStore, TItem>>
+>;
+
+function flatten<TItem>(
```

Comment:
since this is run on every render, can we make this use a for loop instead of reduce? That will be more efficient (i.e. fewer intermediate objects)

In addition, can we simplify what's going on with the `pageInfo`? That seems to come from the last item, always, so can we just do `arr[arr.length - 1]?.pageInfo ?? { hasNextPage: true, endCursor: null }` where necessary?